### PR TITLE
Upgrade Test

### PR DESCRIPTION
**Pull request by DomFleischmann**
Add upgrade test which deploys full bundle in stable channel and tries to upgrade to the bundle in edge.

This test is executed with pytest operator and tox.

Also change github actions to use charmcraft snap instead of pip package.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[tox]
+skipsdist = True
+envlist = upgrade
+
+[testenv]
+setenv =
+    PYTHONPATH={toxinidir}:{toxinidir}/src
+    PYTHONBREAKPOINT=ipdb.set_trace
```
Comment (knkski) [inline tox.ini:8]: This seems like it will interfere with CI if a traceback occurs. Will the process just hang?

File: tests/upgrade/test_upgrade_charm.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import logging
```
Comment (knkski) [inline tests/upgrade/test_upgrade_charm.py:1]: I think we can leave the tests directory flat, i.e. move this up to be under `tests/`. Since pytest offers pretty powerful filtering via `-k`, I'm not worried about organizing the tests themselves into subdirectories.

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+[testenv:upgrade]
+deps =
+    juju
+    pytest
+    pytest-operator
+commands = pytest -v --tb native --show-capture=no --log-cli-level=INFO -s {posargs} {toxinidir}/tests/upgrade
```
Comment (knkski) [inline tox.ini:15]: It seems like if we're going to use tox, we should really adopt it, and have github actions just run `tox`, and have this look something like:

    pytest -k 'not upgrade'
    pytest -k 'upgrade'
    pytest -k 'not upgrade'

That will ensure that all of the regular tests work both before and after an upgrade

File: tests/upgrade/test_upgrade_charm.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    res = subprocess.run(["charm", "-v"])
+    if res.returncode != 0:
+        logging.error(
+            "Charm snap is not installed, please install with: \n \
+                sudo snap install charm --classic"
+        )
+
+    temp_dir = f"{tempfile.mkdtemp()}/kubeflow-bundle"
+
+    res = subprocess.run(["charm", "pull", "cs:kubeflow", "--channel", "edge", temp_dir])
```
[the remark this answers is not in the document]
Comment (DomFleischmann) [reply tests/upgrade/test_upgrade_charm.py:49]: I don't think edge breaking is an issue, it is perfectly fine if the tests afterwards fail because of some other reason. What we really want to make sure with this is that we aren't breaking our upgrade process unintentionally (e.g relation name change, application name change, schema change...)

File: tox.ini (hunk 3 above)
Comment (DomFleischmann) [reply tox.ini:15]: Does that mean you are okay with using tox for the tests that are executed in every PR/merge?

Comment (knkski) [reply tox.ini:15]: Sure, sounds good to me. We'll probably want a tox env for each bundle size, and also have tox accept extra params that it just passes to pytest, to make it easy to do things like `tox -e full -k selenium`

File: ops_tests/test_upgrade_charm.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            ),
+        ]
+    )
+    # raise_on_error=False due to kubeflow-profiles going to an error state if
+    # kubeflow-dashboard CRDs aren't created yet.
+    await ops_test.model.wait_for_idle(raise_on_error=False)
+
+
+async def test_upgrade_apps_in_bundle(ops_test):
+    res = subprocess.run(["charm", "-v"])
```
Comment (knkski) [inline ops_tests/test_upgrade_charm.py:40]: Can we include `charm` in the list of pip packages to install? That way tox will install it automatically

File: ops_tests/test_upgrade_charm.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    if res.returncode != 0:
+        logging.error(
+            "Charm snap is not installed, please install with: \n \
+                sudo snap install charm --classic"
+        )
+
+    temp_dir = f"{tempfile.mkdtemp()}/kubeflow-bundle"
```
Comment (knkski) [inline ops_tests/test_upgrade_charm.py:47]: This seems like it will leave old directories around. Can we run it like this?

```python
with tempfile.TemporaryDirectory() as temp_dir:
    ...
```

File: ops_tests/test_upgrade_charm.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            else:
+                raise error
+    await ops_test.model.wait_for_idle()
+
+
+async def test_status(ops_test):
+    try:
+        await ops_test.model.block_until(
+            lambda: all(
+                (unit.workload_status in ["active"]) and unit.agent_status == "idle"
```
Comment (knkski) [inline ops_tests/test_upgrade_charm.py:70]: Is there a reason this isn't `unit.workload_status == "active"`?

File: ops_tests/test_upgrade_charm.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                for _, application in ops_test.model.applications.items()
+                for unit in application.units
+            ),
+            timeout=600,
+        )
+    except asyncio.TimeoutError:
+        assert False
+    assert True
```
Comment (knkski) [inline ops_tests/test_upgrade_charm.py:78]: Not sure what this is doing

File: ops_tests/test_upgrade_charm.py (hunk 5 above)
Comment (DomFleischmann) [reply ops_tests/test_upgrade_charm.py:40]: pytest-operator expects the charm snap to be installed and will error out if it's not. I tested it with only the pip package and seems to specifically expect the snap. So my check is probably unnecessary anyway.

File: ops_tests/test_upgrade_charm.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        await ops_test.model.block_until(
+            lambda: all(
+                (unit.workload_status == "active") and unit.agent_status == "idle"
+                for _, application in ops_test.model.applications.items()
+                for unit in application.units
+            ),
+            timeout=600,
+        )
+    except asyncio.TimeoutError:
+        assert False
```
Comment (knkski) [inline ops_tests/test_upgrade_charm.py:78]: I assume the `try`/`catch` block here is to make the timeout exception a little easier to read. If so, an `assert False` probably isn't that much more readable. Doing `assert False, 'Some explanatory message'` would work fine, though

**Review by knkski — APPROVED** [review]
small nit, otherwise LGTM
